drive-stack/conUDS/src/main.rs

```diff
@@ -46,68 +50,220 @@ async fn main() {
+                                        node, node_crc, app_crc
+                                    );
+                                }
+                            } else {
+                                error!("Invalid CRC length from node '{}'", node);
+                            }
+                        } else {
+                            error!("Unexpected DID length from node '{}': {:?}", node, node_crc_bytes);
+                        }
+                    }
+                    Err(e) => {
+                        error!("DID read (CRC) failed on node '{}': {:?}", node, e);
+                        // proceed with download anyway
+                    }
+                }
+
+                uds.client.start_persistent_tp().await;
+                let _ = uds.reset_node(SupportedResetTypes::Hard).await;
+
+                if let Err(e) = uds.file_download(&bin, 0x08002000).await {
+                    error!("Download failed for node '{}': {}", node, e);
+                }
+
+                uds.teardown().await;
+            }
+        }
+
+        // ===== single-node commands remain the same, but require -n =====
         ArgSubCommands::Reset(reset) => {
-            debug!(
-                "Performing {:#?} reset for node `{}`",
-                reset.reset_type, args.node
-            );
+            let node = args.node.clone().unwrap_or_else(|| {
+                error!("-n <node> is required for 'reset' (omit only when using 'batch').");
+                std::process::exit(1)
+            });
+            let uds_node = cfg.nodes.get(&node).unwrap_or_else(|| {
+                error!("UDS node '{}' not defined", node);
+                std::process::exit(1)
+            });
+
+            let mut uds = UdsSession::new(&args.device, uds_node.request_id, uds_node.response_id, true).await;
+            info!("Performing {:?} reset for node '{}'", reset.reset_type, node);
             if let Err(e) = uds.reset_node(reset.reset_type).await {
                 error!("Error while resetting ecu: {}", e);
             }
+            uds.teardown().await;
         }
+
         ArgSubCommands::Download(dl) => {
-            debug!(
-                "Downloading binary at '{:#?}' to node `{}`",
-                dl.binary, args.node
-            );
+            let node = args.node.clone().unwrap_or_else(|| {
+                error!("-n <node> is required for 'download' (omit only when using 'batch').");
+                std::process::exit(1)
+            });
+            let uds_node = cfg.nodes.get(&node).unwrap_or_else(|| {
+                error!("UDS node '{}' not defined", node);
+                std::process::exit(1)
+            });
+
+            let mut uds = UdsSession::new(&args.device, uds_node.request_id, uds_node.response_id, args.fast).await;
+            info!("Downloading binary {:?} to node '{}'", dl.binary, node);
+
+            if dl.verify_crc {
```

Comment:
this seems like mostly copy-pasta from batch, maybe write a function for this?

Reply:
Yes need a bit of clean up here